python/engine/loop.py

```
@@ -1779,6 +1921,17 @@ def _softplus(x: float) -> float:
                 _level_start_cursor = (active_loader.cursor()["position"]
                                        if active_loader is not None else 0)
+            # Re-carve window-local val for the new data window
+            if bcfg.window_local_val and active_loader is not None:
+                cursor_pos = active_loader.cursor()["position"]
+                _window_val_tokens, _window_val_targets = _update_window_val(
+                    active_loader, cursor_pos, bcfg.window_val_tokens)
+                if _window_val_tokens is not None:
+                    print(f"  Window-local val: re-carved {len(_window_val_tokens):,} "
+                          f"tokens from new window")
+                else:
+                    print(f"  Window-local val: not enough data remaining, eval disabled")
```

Ruff (0.15.5)

[error] f-string without any placeholders

Remove extraneous `f` prefix

(F541)